### PR TITLE
Update to arkworks 0.4.0

### PR DESCRIPTION
**Pull request by achimcc**
This updates all Arkworks dependencies to the most recent Arkworks 0.4.0/0.4.2 versions.

Solves issue https://github.com/nikkolasg/snarkpack/issues/9

### CODE REVIEW

File: src/ip.rs [hunk 1]
```diff
@@ -1,43 +1,44 @@
...
+    AffineRepr, VariableBaseMSM,
+};
+use ark_std::vec::Vec;
 
-pub(crate) fn pairing_miller_affine<E: PairingEngine>(
+pub(crate) fn pairing_miller_affine<E: Pairing>(
     left: &[E::G1Affine],
     right: &[E::G2Affine],
-) -> Result<E::Fqk, Error> {
+) -> Option<MillerLoopOutput<E>> {
```
Comment (nikkolasg) [inline src/ip.rs:11]: Can we keep that a result to make it similar to the other functions there ?

**Review by nikkolasg — COMMENTED** [review]
That's great, thanks a lot for this effort !
I have mainly two requests at this point before merging, let me know how that sounds:
* Could you run a quick performance test, manually, before and after to see how perf evolves with 0.4 ?
* Would you be willing to go the extra miles and add support for CI, just at least testing ?

I can help on those as well if it's too much, let me know how that sounds.

Thank you !

File: src/ip.rs (hunk 1 above)
Comment (achimcc) [reply src/ip.rs]: Yes, sure! Makes sense, done!

**Comment by achimcc** [comment]
> That's great, thanks a lot for this effort ! I have mainly two requests at this point before merging, let me know how that sounds:
> 
> * Could you run a quick performance test, manually, before and after to see how perf evolves with 0.4 ?
> * Would you be willing to go the extra miles and add support for github action CI, just at least testing, so we can see that this PR doesn't break anything before merging ? There's already circle CI but now I think github action is just strictly simpler and better.
> 
> I can help on those as well if it's too much, let me know how that sounds.
> 
> Thank you !

Let me do another PR first tomorrow, which adds basic CI and benchmarking to the status quo, before we merge this PR. I will also share some bench results here tomorrow!

**Comment by achimcc** [comment]
@nikkolasg : I created this PR, which we can merge first, it also adds a bot which reports benchmark diffs of future PR's:

https://github.com/nikkolasg/snarkpack/pull/10

**Comment by achimcc** [comment]
These are the criterion benchmarking results I get from running the `groth16_aggregation` on my local machine (Core i7-10610U CPU @ 1.80GHz, 16GB RAM):

Arkworks 0.3.0:

groth16 aggregation     time:   [1.0761 s **1.0787 s** 1.0814 s]

Arkworks 0.4.0:

groth16 aggregation     time:   [882.16 ms **884.54 ms** 887.10 ms]

**Comment by nikkolasg** [comment]
Awesome - seems like passing to 0.4 improved by 10%.

BTW if you're using this / plan to using this in prod, we can talk more down the line there are further math optimization to be done.
Also, I can also give you a maintainer role in the repo if you're likely to hack on it often. Just let me know.

Happy to merge this way once conflict resolved.

Thanks a lot and thanks for the CI PR !

**Comment by achimcc** [comment]
Sure! We can also talk about adding me as a maintainer!

We are not aiming for using this in production immediately, but perhaps in the future.

I'd recommend removing the CIRCLE CI then, since it seems not to work any longer?

If you create a `PUBLISH_TOKEN` in the repos secrets first, we will get the benchmark diffs posted here ;)

**Comment by achimcc** [comment]
I did some of the clippy suggestions, I'm happy to take another look at the remaining ones tomorrow!

**Comment by achimcc** [comment]
There are so many suggested clippy fixes, that I reverted the ones that I already solved. I think for reviewability; it is better to do a follow-up PR which addresses them, where I cherry pick them from this PR.

**Comment by nikkolasg** [comment]
I've made you a collaborator - need to accept invite.

I've setup a secret as well named PUBLISH_TOKEN and containing an API publish token for this repo but the GH action doesn't seem to pick it up, it's weird (I've actually never used those before so I may be missing a big step).

**Comment by nikkolasg** [comment]
Ahah i know why: these are not passed from a PR that comes from a fork ;)

**Comment by nikkolasg** [comment]
I'll merge and then we can do another PR from this repo to trigger it.

**Pull request merged**